lib/patches/bundler/compact_index_client/updater.rb

```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
+require "fileutils"
+require "stringio"
+require "tmpdir"
+require "zlib"
+
+module Bundler
+  class CompactIndexClient
+    class Updater
+      def update(local_path, remote_path, retrying = nil)
+        headers = {}
+
+        Dir.mktmpdir("bundler-compact-index-") do |local_temp_dir|
+          local_temp_path = Pathname.new(local_temp_dir).join(local_path.basename)
+
+          # first try to fetch any new bytes on the existing file
+          if retrying.nil? && local_path.file?
+            FileUtils.cp local_path, local_temp_path
+            headers["If-None-Match"] = etag_for(local_temp_path)
+            headers["Range"] = "bytes=#{local_temp_path.size}-"
+          else
+            # Fastly ignores Range when Accept-Encoding: gzip is set
+            headers["Accept-Encoding"] = "gzip"
+          end
+
+          response = @fetcher.call(remote_path, headers)
+          return nil if response.is_a?(Net::HTTPNotModified)
+
+          content = response.body
+          if response["Content-Encoding"] == "gzip"
+            # content = Zlib::GzipReader.new(StringIO.new(content)).read
+            IO.binwrite("#{local_temp_dir}/gzipped_versions", content)
+            content = `gunzip -c #{local_temp_dir}/gzipped_versions`
```

Comment:
Not for this PR, but I think we should have this workaround enabled just on the `zlib` require.
Then we don't need to replicate large amounts of code and it works for other usages.